### PR TITLE
Fix #77

### LINKED ISSUE #77: [4.1.0] Timer countdown stops when screen is off
OS: Lineage OS 14.1

This started happening with 4.1.0. It worked flawlessly with 4.0.1 (both versions from fdroid).

I have a timer of 38 seconds. Vibrate is on. No sound.
I start it and I turn the screen off. With 4.0.1 the timer continued running and when it expired it started vibrating and the screen lit up allowing me to stop it.
Now with 4.1.0, the timer seems to stop the countdown with about 20 seconds remaining when the screen is off. When I turn the screen on and unlock it, the timer countdown continues.
If I leave the screen on from the start the problem doesn't exist.

I verified this by going back to 4.0.1 and it continues to work.
With 4.1.0 I even disabled battery optimization but it still didn't work.

### PR DESCRIPTION
**Pull request by pavolfranekdev**
Fixed issue #77 Do not stop the timer at closing the app

### CODE REVIEW

File: app/src/main/kotlin/com/simplemobiletools/clock/services/timerService.kt [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+
+    @Subscribe(threadMode = ThreadMode.MAIN)
+    fun onMessageEvent(state: TimerState.Paused) {
+        config.timerState = state
+        timer?.cancel()
+        stopService()
+    }
+
+    private fun stopService() {
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) stopForeground(true)
```
Comment (tibbi) [inline app/src/main/kotlin/com/simplemobiletools/clock/services/timerService.kt]: missing brackets, split it into multiple lines. There is also an isOreoPlus helper function, use that

File: app/src/main/kotlin/com/simplemobiletools/clock/services/timerService.kt [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+import com.simplemobiletools.clock.extensions.secondsToMillis
+import com.simplemobiletools.clock.helpers.TIMER_NOTIF_ID
+import com.simplemobiletools.clock.helpers.TIMER_RUNNING_NOTIF_ID
+import com.simplemobiletools.commons.extensions.getFormattedDuration
+import com.simplemobiletools.commons.helpers.isOreoPlus
+import org.greenrobot.eventbus.EventBus
+import org.greenrobot.eventbus.Subscribe
+import org.greenrobot.eventbus.ThreadMode
+
+fun Context.startTimerService() {
```
Comment (tibbi) [inline app/src/main/kotlin/com/simplemobiletools/clock/services/timerService.kt]: move it into the context extensions to have them in one place

File: app/src/main/kotlin/com/simplemobiletools/clock/extensions/gson/typeAdapter.kt [hunk 3]
```diff
@@ -0,0 +1,22 @@
+package com.simplemobiletools.clock.extensions.gson
```
Comment (tibbi) [inline app/src/main/kotlin/com/simplemobiletools/clock/extensions/gson/typeAdapter.kt]: use capital T at "typeAdapter.kt" filename

File: app/src/main/kotlin/com/simplemobiletools/clock/fragments/TimerFragment.kt [hunk 4]
```diff
@@ -81,208 +84,72 @@ class TimerFragment : Fragment() {
...
-            if (context?.isScreenOn() == true) {
-                context!!.showTimerNotification(false)
-                Handler().postDelayed({
-                    context?.hideTimerNotification()
-                }, context?.config!!.timerMaxReminderSecs * 1000L)
-            } else {
-                Intent(context, ReminderActivity::class.java).apply {
-                    activity?.startActivity(this)
+        val drawableId =
+                if (timerState is TimerState.Running) {
```
Comment (tibbi) [inline app/src/main/kotlin/com/simplemobiletools/clock/fragments/TimerFragment.kt]: add this to the previous line, like "val drawableId = if (timerState..."

**Comment by tibbi** [comment]
lets show the notification only if the app gets closed while the timer runs, like in Google Clock

**Comment by tibbi** [comment]
If I set the timer to 7 seconds and start it, it starts from 6
If I keep the app foregrounded and the timer runs out, the Reset button disappears and I have to use the notification to stop it. As previously written, the notification shouldnt even be there if the app is in the foreground.

File: app/src/main/kotlin/com/simplemobiletools/clock/services/timerService.kt [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+
+    override fun onBind(intent: Intent?): IBinder? = null
+
+    override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int {
+        super.onStartCommand(intent, flags, startId)
+
+        val formattedDuration = config.timerSeconds.getFormattedDuration()
+        startForeground(TIMER_RUNNING_NOTIF_ID, notification(formattedDuration))
+
+        when (val state = config.timerState) {
```
Comment (tibbi) [inline app/src/main/kotlin/com/simplemobiletools/clock/services/timerService.kt]: these switch/case lines are quite confusing, simplify it please

File: app/src/main/kotlin/com/simplemobiletools/clock/services/timerService.kt (hunk 2 above)
Comment (pavolfranekdev) [reply app/src/main/kotlin/com/simplemobiletools/clock/services/timerService.kt]: this is not generally extension. It is not applicable again to other conditions/services and therefore I would not move it. This function is directly related to TimerService and therefore it is convenient when everything is in one place.

Comment (tibbi) [reply app/src/main/kotlin/com/simplemobiletools/clock/services/timerService.kt]: if it is related to this only, make it a casual function. Extensions are not created for these cases.

File: app/src/main/kotlin/com/simplemobiletools/clock/services/timerService.kt (hunk 5 above)
Comment (pavolfranekdev) [reply app/src/main/kotlin/com/simplemobiletools/clock/services/timerService.kt]: start from 6 fixed. Replaced toInt with roundToInt

**Comment by tibbi** [comment]
looks good, thanks

**Pull request merged**